lib/krb5/krcache.c

```diff
@@ -0,0 +1,1972 @@
+    ret = (key == -1) ? errno : 0;
+    krb5_data_free(&payload);
+
+cleanup:
+    krb5_storage_free(sp);
+
+    return ret;
+}
+
+static krb5_error_code
+parse_index(krb5_context context, int32_t *version, char **primary,
+	    const unsigned char *payload, size_t psize)
+{
+    krb5_error_code ret;
+    krb5_data payload_data;
+    krb5_storage *sp;
+
+    payload_data.length = psize;
+    payload_data.data = rk_UNCONST(payload);
+
+    sp = krb5_storage_from_data(&payload_data);
+    if (sp == NULL)
+	return KRB5_CC_NOMEM;
+
+    krb5_storage_set_byteorder(sp, KRB5_STORAGE_BYTEORDER_BE);
+
+    ret = krb5_ret_int32(sp, version);   
+    if (ret == 0)
+	ret = krb5_ret_string(sp, primary);
+
+    krb5_storage_free(sp);
+
+    return ret;
+}
+
+/*
+ * Get or initialize the primary name within collection_id and set
+ * *subsidiary_out to its value.  If initializing a legacy collection, look
+ * for a legacy cache and add it to the collection.
+ */
+static krb5_error_code
+get_primary_name(krb5_context context,
+		 const char *anchor_name,
+		 const char *collection_name,
+		 key_serial_t collection_id,
+		 char **subsidiary_out)
+{
+    krb5_error_code ret;
+    key_serial_t primary_id, legacy;
+    void *payload = NULL;
+    int payloadlen;
+    int32_t version;
+    char *subsidiary_name = NULL;
+
+    *subsidiary_out = NULL;
+
+    primary_id = keyctl_search(collection_id, KRCC_KEY_TYPE_USER,
+			       KRCC_COLLECTION_PRIMARY, 0);
+    if (primary_id == -1) {
+	/* Initialize the primary key using the collection name.  We can't name
+	 * a key with the empty string, so map that to an arbitrary string. */
+	subsidiary_name = strdup((*collection_name == '\0') ? "tkt" :
+				 collection_name);
+	if (subsidiary_name == NULL) {
+	    ret = krb5_enomem(context);
+	    goto cleanup;
+	}
+	ret = set_primary_name(context, collection_id, subsidiary_name);
+	if (ret)
+	    goto cleanup;
+
+	if (strcmp(anchor_name, KRCC_LEGACY_ANCHOR) == 0) {
+	    /* Look for a cache created by old code.  If we find one, add it to
+	     * the collection. */
+	    legacy = keyctl_search(KEY_SPEC_SESSION_KEYRING,
+				   KRCC_KEY_TYPE_KEYRING, subsidiary_name, 0);
+	    if (legacy != -1 && keyctl_link(legacy, collection_id) == -1) {
+		ret = errno;
+		goto cleanup;
+	    }
+	}
+    } else {
+	/* Read, parse, and free the primary key's payload. */
+	payloadlen = keyctl_read_alloc(primary_id, &payload);
+	if (payloadlen == -1) {
+	    ret = errno;
+	    goto cleanup;
+	}
+	ret = parse_index(context, &version, &subsidiary_name, payload,
+			  payloadlen);
+	if (ret)
+	    goto cleanup;
+
+	if (version != KRCC_COLLECTION_VERSION) {
+	    ret = KRB5_KCC_UNKNOWN_VERSION;
+	    goto cleanup;
+	}
+    }
+
+    *subsidiary_out = subsidiary_name;
+    subsidiary_name = NULL;
+
+cleanup:
+    free(payload);
+    free(subsidiary_name);
+
+    return ret;
+}
+
+/*
+ * Note: MIT keyring code uses krb5int_random_string() as if the second argument
```

Comment:
Hmm?  Near as I can tell the MIT keyring code passes `KRCC_NAME_RAND_CHARS` (which is 8) as the `slen` argument, and that's taken to be a buffer size in bytes, with the last being set to `'\0'`.

Reply:
But the containing variable (from memory) is sizeof(prefix) + KRCC_NAME_RAND_CHARS which suggests to me that it’s a character count not a buffer size.

Reply:
I'm pretty sure they mean "CHARS" as in "char", not as in "characters".  There's no string yet whose Unicode (or whatever) characters to count -- it can only be bytes.

Reply:
BTW, thanks for this comment, it made me look more closely.